src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+                document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(document);
+            }
+
+            try
+            {
+                var extension = _extensions.GetOrAdd(assemblyPath, _ =>
+                {
+                    var directory = Path.GetDirectoryName(assemblyPath);
+                    Contract.ThrowIfNull(directory);
+
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyPath}", isCollectible: true);
+                    loadContext.Resolving += (context, assemblyName) =>
+                    {
+                        var sharedAssembly = _defaultLoadContext.Assemblies.Where(a => a.GetName().Name == assemblyName.Name).FirstOrDefault();
+
+                        if (sharedAssembly is not null)
+                        {
+                            if (sharedAssembly.GetName().Version < assemblyName.Version)
+                            {
+                                throw new InvalidOperationException($"The version of the loaded assembly {assemblyName.Name} is too low: requested {assemblyName.Version}, found {sharedAssembly.GetName().Version}.");
+                            }
+
+                            return sharedAssembly;
+                        }
+
+                        var extensionAssemblyPath = Path.Combine(directory, $"{assemblyName.Name}.dll");
+                        if (File.Exists(extensionAssemblyPath))
+                        {
+                            return loadContext.LoadFromAssemblyPath(extensionAssemblyPath);
+                        }
+
+                        return null;
+                    };
+
+                    return new CustomMessageHandlerExtension(loadContext);
+                });
+
+                var handler = extension.Handlers.GetOrAdd(typeFullName, name =>
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+                    var type = assembly.GetType(typeFullName)
+                        ?? throw new InvalidOperationException($"Cannot find type {typeFullName} in {assemblyPath}.");
+
+                    // Create the Handler instance. Requires having a parameterless constructor.
+                    // ```
+                    // public class CustomMessageHandler
+                    // {
+                    //     public Task<TResponse> ExecuteAsync(TRequest, CancellationToken);
+                    // }
+                    //
+                    // public class CustomMessageDocumentHandler
+                    // {
+                    //     public Task<TResponse> ExecuteAsync(TRequest, Document, CancellationToken);
+                    // }
+                    // ```
+                    return Activator.CreateInstance(type)
```

Comment:
similar to before, I think it is possible that this gets called multiple times (and create multiple instances).  We might want more explicit locks to ensure we're only creating once instance (say if the plugin construction has side effects)